### DISCUSSION TITLE
[Other] Improve running Paperless-ng with a custom database port

### DISCUSSION THREAD

**Discussion by holzhannes** [jonaswinkler/paperless-ng#730] [Support]
I am running paperless-ng on my Synology-NAS with a custom port for the PostgreSQL Database. Therefore, I need to change the value of `PAPERLESS_DBHOST` from `db` to the ip address of the network gateway for example: `172.18.0.1` displayed in the Synology Docker UI. I am not sure, but maybe it would be possible to improve the `docker-compose.yml` to not manual edit the PAPERLESS_DBHOST. Maybe one of you has an idea.

Here are my files I am using:

## docker-compose.env
```
# docker-compose.env file for running paperless from the Docker Hub on a Synology Nas with a custom database port

USERMAP_UID=1026
USERMAP_GID=100

PAPERLESS_OCR_LANGUAGES=deu eng
PAPERLESS_OCR_LANGUAGE=deu

PAPERLESS_SECRET_KEY=SECRET
PAPERLESS_TIME_ZONE=Europe/Berlin

PAPERLESS_DBPORT=5434
```

## docker-compose.yml

```
# docker-compose file for running paperless from the Docker Hub on a Synology Nas with a custom database port

version: "3.4"
services:
  broker:
    image: redis:6.0
    restart: unless-stopped

  db:
    image: postgres:13
    restart: unless-stopped
    volumes:
      - /volume1/Paperless/db:/var/lib/postgresql/data
    environment:
      POSTGRES_DB: paperless
      POSTGRES_USER: paperless
      POSTGRES_PASSWORD: paperless
    ports:
      - 5434:5432

  webserver:
    image: jonaswinkler/paperless-ng:latest
    restart: unless-stopped
    depends_on:
      - db
      - broker
      - gotenberg
      - tika
    ports:
      - 8082:8000
    healthcheck:
      test: ["CMD", "curl", "-f", "http://localhost:8000"]
      interval: 30s
      timeout: 10s
      retries: 5
    volumes:
      - /volume1/Paperless/data:/usr/src/paperless/data
      - /volume1/Paperless/media:/usr/src/paperless/media
      - /volume1/Paperless/export:/usr/src/paperless/export
      - /volume1/Scan2Paperless:/usr/src/paperless/consume
    env_file: docker-compose.env
    environment:
      PAPERLESS_REDIS: redis://broker:6379
      PAPERLESS_DBHOST: 172.18.0.1
      PAPERLESS_TIKA_ENABLED: 1
      PAPERLESS_TIKA_GOTENBERG_ENDPOINT: http://gotenberg:3000
      PAPERLESS_TIKA_ENDPOINT: http://tika:9998

  gotenberg:
    image: thecodingmachine/gotenberg
    restart: unless-stopped
    environment:
      DISABLE_GOOGLE_CHROME: 1

  tika:
    image: apache/tika
    restart: unless-stopped

volumes:
  data:
  media:
  pgdata: 
 ```
Thanks a lot to @jonaswinkler for your great and effort for paperless-ng! I previously used Mayan EDMS, which is also great, but a little too complex for my use case.

**Comment by jonaswinkler** [comment]
I'm not entirely sure what you're trying to achieve.

> I am running paperless-ng on my Synology-NAS with a custom port for the PostgreSQL Database.

If you need to adjust the port, there's `PAPERLESS_DBPORT` for this.

Are you already running a PostgreSQL server and want paperless to use that? If so, you should remove the `db:` service from the compose file entirely - this defines and starts an entire PostgreSQL server on its own, using the default port 5432.  `PAPERLESS_DBHOST: db` tells paperless to use that server.

Also, you probably need to fiddle with docker networking settings a bit. By default, the docker-compose file tells docker to create a private network just for paperless, and runs all containers (gotenberg, tika, database, webserver, broker) in that network. They can reach each other, but nothing else.

The docker-compose file is really just a template that works out of the box, but as soon as you want to do something custom, you *need* to edit it.

**Comment by holzhannes** [comment]
> I'm not entirely sure what you're trying to achieve.
> 
> > I am running paperless-ng on my Synology-NAS with a custom port for the PostgreSQL Database.
> 
> If you need to adjust the port, there's `PAPERLESS_DBPORT` for this.
> 
> Are you already running a PostgreSQL server and want paperless to use that?

Sorry I think my explanation was not clear enough. I am already running a PostgreSQL Server on Port 5432 (used by Mayan EDMS), but I don't like to use it. I like to use the separate PostgreSQL server created by the compose file with Port 5434 in my example. Therefore, I added 

```    
ports:
      - 5434:5432
```
and changed `PAPERLESS_DBHOST: db` to `PAPERLESS_DBHOST: 172.18.0.1` because this is the only way it was working. The only thing I was thinking about was if there is a solution to **not** fiddle around with the IP. 

Do you think my Setup is strange and it is better to use only one PostgreSQL Server?

**Comment by jonaswinkler** [comment]
Well, you can do anything you like, but I'd probably use one PostgreSQL server. These can handle multiple databases of multiple applications.

You might want to read up on networking with docker compose: https://docs.docker.com/compose/networking/.

* If you want paperless to use the postgresql server defined in the compose file, you don't need to change anything, not even the port.
* Also `ports:` simply exposes that port outside of the default network compose creates for you, but this is probably not something you want.
* If you want paperless to use an existing PostgreSQL server, you *need* to fiddle with the host and port settings.
